Add hands-on workshop for Moderne agent tools (#636)
moderneinc/moderne-docs: docs/hands-on-learning/agent-tools/module-1-cli-and-lsts.md

---
sidebar_label: "Module 1: CLI and LSTs"
description: Install the Moderne CLI, install the Prethink recipe modules, and build LSTs for a working set of Java and Spring projects.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Module 1: CLI and LSTs

> In this module, you'll set up everything the agent tools depend on: the Moderne CLI, a couple of recipe JARs (including `io.moderne.recipe:rewrite-prethink`), and Lossless Semantic Trees (LSTs) for a small working set of Java and Spring repositories. The later modules build on top of this setup, so don't skip ahead until your LSTs build cleanly.

Reviewer: Minor - but I think it would be nice to link to the LST doc as this is the first module: https://docs.moderne.io/administrator-documentation/moderne-platform/references/lossless-semantic-trees/

<Tabs groupId="cli-install-os" queryString="os">
<TabItem value="linux-macos" label="Linux / macOS" default>

```bash
curl https://app.moderne.io/cli | bash
```

</TabItem>
<TabItem value="windows" label="Windows">

```powershell
irm https://app.moderne.io/cli/windows | iex
```

</TabItem>
</Tabs>

Alternatively, you can install via a package manager:

* **Homebrew** (macOS/Linux): `brew install moderneinc/moderne/mod`
* **Chocolatey** (Windows): `choco install mod --prerelease`

For more installation options, including Enterprise and Moderne DX setups, see [Getting started with the Moderne CLI](../../user-documentation/moderne-cli/getting-started/cli-intro.md#installation-and-configuration).

Verify the install:

```bash
mod --version
```

#### Step 2: Connect the CLI to Moderne

Point the CLI at the public Moderne tenant and log in:

```bash
mod config moderne edit https://app.moderne.io
mod config moderne login
```

The login command opens a browser, asks you to grant the CLI access to your account, and stores a token locally that's valid for 365 days. If you're running on an Enterprise tenant, replace the URL with your tenant's hostname.

:::tip Restrictive networks

> On corporate networks with proxies or limited Maven Central / Moderne SaaS access, see [Using the CLI with internal tools and artifact repositories](../../user-documentation/moderne-cli/getting-started/cli-internal-tools.md) before continuing. Symptoms you might hit later in this workshop:

Reviewer: ```suggestion
On corporate networks with proxies or limited Maven Central / Moderne SaaS access, see [Using the CLI with internal tools and artifact repositories](../../user-documentation/moderne-cli/getting-started/cli-internal-tools.md) before continuing. Without a properly configured setup, you'll commonly see:
```

* `mod git sync moderne` reports `PARTIAL SUCCESS` because a few LSTs sit on hosts your network can't reach
* `mod config recipes jar install` fails to resolve a JAR — point it at your internal Artifactory/Nexus first via `mod config recipes artifacts`
* `mod build` can't download Maven/Gradle plugins — set up your build tool's mirror and proxy as you normally would, then re-run

> For the workshop you only need a handful of repos to succeed. Don't try to fix every failure up front.

Reviewer: ```suggestion
For the workshop you only need a handful of repos to succeed. **Don't try to fix every failure up front**.
```

:::

### Takeaways

> * The Moderne CLI is the entry point for everything in this workshop. Skills, the MCP server, Prethink, and Trigrep all run through `mod`.

Reviewer: I find it a bit strange we are calling out skills, MCP, Prethink, etc - as those aren't mentioned at all prior to this. 

Perhaps we can say something like this instead?

```suggestion
* The Moderne CLI is the entry point for everything in this workshop. Every command you’ll run starts with `mod`.
```

* You only need to authenticate once. The CLI caches your token across terminal sessions.

---

## Exercise 1-2: Install the recipe modules

### Goals for this exercise

* Install the recipe JARs needed for Prethink and the rest of this workshop
* Understand the difference between `mod config recipes jar install` and a full catalog sync

### Steps

#### Step 1: Install the workshop recipe JARs

Instead of syncing the entire recipe catalog, install just the JARs you need:

<Tabs groupId="cli-install-os" queryString="os">
<TabItem value="linux-macos" label="Linux / macOS" default>

```bash
mod config recipes jar install \

>   org.openrewrite.recipe:rewrite-prethink \

Author: ```suggestion
```

  io.moderne.recipe:rewrite-prethink \
  org.openrewrite.recipe:rewrite-static-analysis \
  org.openrewrite.recipe:rewrite-spring
```

</TabItem>
<TabItem value="windows" label="Windows">

```powershell
mod config recipes jar install `

>   org.openrewrite.recipe:rewrite-prethink `

Author: ```suggestion
```

  io.moderne.recipe:rewrite-prethink `
  org.openrewrite.recipe:rewrite-static-analysis `
  org.openrewrite.recipe:rewrite-spring
```

</TabItem>
</Tabs>

> The two `rewrite-prethink` artifacts are the focus of this workshop: the `org.openrewrite.recipe` module is the open-source building blocks, and `io.moderne.recipe:rewrite-prethink` adds pre-configured discovery for Spring, JPA, Kafka, and other common frameworks. See [Recipe modules](../../user-documentation/agent-tools/prethink.md#recipe-modules) for the full breakdown.

Author: ```suggestion
The `rewrite-prethink` artifact is the focus of this workshop: it uses an open-source `org.openrewrite.recipe` module for building blocks, while `io.moderne.recipe:rewrite-prethink` adds pre-configured discovery for Spring, JPA, Kafka, and other common frameworks. See [Recipe modules](../../user-documentation/agent-tools/prethink.md#recipe-modules) for the full breakdown.
```
